aioxmpp/avatar/service.py

```diff
@@ -462,39 +641,69 @@ def get_avatar_metadata(self, jid, *, require_fresh=False):
         an empty list of avatar descriptors, since this is
         semantically equivalent to not having an avatar.
         """
-        if not require_fresh:
-            try:
-                return self._metadata_cache[jid]
-            except KeyError:
-                pass
-
-        with (yield from self._notify_lock):
-            if jid in self._metadata_cache:
-                if require_fresh:
-                    del self._metadata_cache[jid]
-                else:
-                    return self._metadata_cache[jid]
 
+        def try_vcard_fallback():
             try:
-                metadata_raw = yield from self._pubsub.get_items(
-                    jid,
-                    namespaces.xep0084_metadata,
-                    max_items=1
-                )
-            except aioxmpp.XMPPCancelError as e:
-                # transparently map feature-not-implemente and
-                # item-not-found to be equivalent unset avatar
-                if e.condition in (
-                        (namespaces.stanzas, "feature-not-implemented"),
-                        (namespaces.stanzas, "item-not-found")):
+                logger.debug("trying vCard avatar as fallback for %s", jid)
+                vcard = yield from self._vcard.get_vcard(jid)
+                # XXX: get_photo_data should extract the MIME type as well
+                # here we can know, so we should pass it on
+                photo = vcard.get_photo_data()
+                if photo is not None:
+                    logger.debug("success vCard avatar as fallback for %s", jid)
+                    sha1 = hashlib.sha1()
+                    sha1.update(photo)
                     metadata = collections.defaultdict(lambda: [])
-                else:
-                    raise
+                    metadata[None] = [VCardAvatarDescriptor(
+                        remote_jid=jid,
+                        mime_type=None,
+                        id_=normalize_id(sha1.hexdigest()),
+                        vcard=self._vcard,
+                        nbytes=len(photo),
+                        image_bytes=photo,
+                    )]
+                    return metadata
+            except aioxmpp.XMPPCancelError:
```

Comment:
Make the ``try`` block as small as possible and as large as necessary. The ``logger`` call and everything after ``get_vcard`` should be outside of the ``try`` block.